src/plugin_builder.rs

```diff
@@ -0,0 +1,55 @@
+use std::{marker::PhantomData, sync::{Arc, Mutex}};
+use bevy::{app::App, prelude::World};
+
+use crate::{callback::{Callback, IntoCallbackSystem}, Callbacks, Runtime, ScriptingRuntimeBuilder};
+
+pub struct ApiBuilder<'a, R: Runtime> {
+    _phantom_data: PhantomData<R>,
+    world: &'a mut World,
+}
+
+impl<'a, R: Runtime> ApiBuilder<'a, R> {
+    fn new(world: &'a mut World) -> Self {
+        Self {
+            _phantom_data: PhantomData,
+            world,
+        }
+    }
+
+    /// Registers a function for calling from within a script.
+    /// Provided function needs to be a valid bevy system and its
+    /// arguments and return value need to be convertible to runtime
+    /// value types.
+    pub fn add_function<In, Out, Marker>(
+        self,
+        name: String,
+        fun: impl IntoCallbackSystem<R, In, Out, Marker>,
+    ) -> Self {
+        let system = fun.into_callback_system(self.world);
+
+        let mut callbacks_resource = self.world.resource_mut::<Callbacks<R>>();
+
+        callbacks_resource.uninitialized_callbacks.push(Callback {
```

Comment:
Since we are doing this in 2 places now, I think it might be beneficial to extract this to something like `Callbacks::add_callback` or similar

Reply:
Gladly! 

I moved the code to the `BuildScriptingRuntime`. Because of that there is also still one use of the `callbacks_resource` code you mentioned, so extraction of that code is not required.

I added a working example for lua and will work on the documentation a bit tomorrow.

Honestly I tried implementing mlua myself and failing miserably at that before I found this crate. It was all a bit above my skill level, saying that I looked into the test script and have to admit I'm not sure how that's working.

So ideally I work on the documentation some more, add a rhai example, work on any comments you may have and then leave writing the test(s) to you ;).

And thank you for this crate!